File: P32_gbeex_v4/gbeex-backend/models/companies.py
# models/companies.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from models.subjects import Subject, VitalSigns, EmergencyContact 

# Remove later if not needed
class CompanyListItem(BaseModel):
    companyId: str
    companyName: str

class ProtocolListItem(BaseModel):
    protocolId: str
    protocolName: str

class SiteLocation(BaseModel):
    lat: float
    lng: float

class SitePerformanceSummary(BaseModel):
    totalTrialsDone: int
    trialSuccessRate: float
    totalSubjectsEnrolledCount: int
    subjectSuccessRate: float
    enrollmentRatePerTrial: float
    dropoutRatePerTrial: float

class SiteDashboardItem(BaseModel):
    siteId: str
    siteName: str
    country: str
    location: SiteLocation
    sitePerformanceSummary: SitePerformanceSummary

# Full Detail Models
# Nested model for Principal Investigator
class PrincipalInvestigator(BaseModel):
    name: str
    credentials: str
    email: str

# Nested model for Contact Info
class ContactInfo(BaseModel):
    name: str
    email: str
    phone: str

# Nested model for Site Performance 
class SiteFullPerformance(BaseModel):
    totalTrialsDone: int
    trialSuccessRate: float
    totalSubjectsEnrolledCount: int
    subjectSuccessRate: float
    enrollmentRatePerTrial: float
    dropoutRatePerTrial: float
    queryResolutionTime: str
    protocolDeviationRate: float
    complicationRate: float

# Nested model for Site Metrics
class SiteMetrics(BaseModel):
    screenFailureRate: float
    avgTimeToContract: str
    avgTimeToIRBApproval: str

# Nested model for IRB Info
class IRBInfo(BaseModel):
    name: str
    registrationNumber: str
    contactEmail: str

# Nested model for Staff Count
class StaffCount(BaseModel):
    doctor: int
    technician: int

# Nested model for Financials 
class SiteFinancials(BaseModel):
    contractValueUSD: int
    paymentsMadeUSD: int

# Full Site Model 
class Site(BaseModel):
    siteId: str
    siteName: str
    siteType: str
    country: str
    state: str
    city: str
    latitude: float
    longitude: float
    principalInvestigator: PrincipalInvestigator
    investigatorExperience: int
    contactInfo: ContactInfo
    siteActivationDate: str 
    recruitmentCapacity: int
    infrastructureScore: float
    sitePerformance: SiteFullPerformance
    siteMetrics: SiteMetrics
    auditHistory: List[str] 
    performanceTrends: str
    timelineDelays: Dict[str, Any] 
    IRBInfo: IRBInfo
    patientRecruitmentSources: List[str]
    equipment: List[str]
    staffCount: StaffCount
    monitoringVisits: int
    financials: SiteFinancials
    subjects: List[Subject] 

# Nested model for Progress Metrics
class ProgressMetrics(BaseModel):
    enrolled: int
    screened: int
    completed: int
    dropped: int
    completionPercentage: float

# Nested model for Ethics Committee Approval
class EthicsCommitteeApproval(BaseModel):
    committeeName: str
    approvalDate: str 
    approvalId: str

# Full Protocol Model 
class Protocol(BaseModel):
    protocolId: str
    protocolName: str
    nctId: str
    drugName: str
    therapeuticArea: str
    phase: str
    status: str
    studyDesign: str
    principalInvestigator: str 
    coordinatingCenter: str
    budgetUSD: int
    startDate: str 
    expectedEndDate: str 
    actualEndDate: str 
    estimatedEnrollmentSite: int
    requiredSiteEnrollment: int
    estimatedEnrollmentSubject: int
    requiredSubjectEnrollment: int
    protocolVersion: str
    lastUpdatePosted: str 
    timelineDelays: Dict[str, Any]
    progressMetrics: ProgressMetrics
    dataMonitoringCommittee: str
    ethicsCommitteeApproval: EthicsCommitteeApproval
    regulatorySubmissions: List[str]
    patientReportedOutcomes: str
    sites: List[Site] 

# Nested model for Financials (Company level)
class CompanyFinancials(BaseModel):
    revenueUSD: int
    rdExpenditureUSD: int
    marketCapUSD: int

# Nested model for Pipeline Summary
class PipelineSummary(BaseModel):
    phase1Count: int
    phase2Count: int
    phase3Count: int

class Company(BaseModel):
    companyId: str
    companyName: str
    sponsorType: str
    financials: CompanyFinancials
    pipelineSummary: PipelineSummary
    activeRegions: List[str]
    therapeuticAreasCovered: List[str]
    complianceScore: float
    riskLevel: str
    website: str
    foundedYear: int
    trialsRan: int
    successRatio: float
    avgTrialCompletionTime: int
    totalSubjectsHandled: int
    protocols: List[Protocol] 

File: P32_gbeex_v4/gbeex-frontend/app/types/companies.ts
// app/types/companies.ts

import type { Subject, VitalSigns, EmergencyContact } from '@/app/types/subjects';
export interface CompanyListItem {
    companyId: string;
    companyName: string;
}

export interface ProtocolListItem {
    protocolId: string;
    protocolName: string;
}

export interface SiteLocation {
    lat: number;
    lng: number;
}

export interface SitePerformanceSummary {
    totalTrialsDone: number;
    trialSuccessRate: number;
    totalSubjectsEnrolledCount: number;
    subjectSuccessRate: number;
    enrollmentRatePerTrial: number;
    dropoutRatePerTrial: number;
}

export interface SiteDashboardItem {
    siteId: string;
    siteName: string;
    country: string;
    location: SiteLocation;
    sitePerformanceSummary: SitePerformanceSummary;
}


// Full Detail Types

// Principal Investigator
export interface PrincipalInvestigator {
    name: string;
    credentials: string;
    email: string;
}

// Contact Info
export interface ContactInfo {
    name: string;
    email: string;
    phone: string;
}

// Site Performance 
export interface SiteFullPerformance {
    totalTrialsDone: number;
    trialSuccessRate: number;
    totalSubjectsEnrolledCount: number;
    subjectSuccessRate: number;
    enrollmentRatePerTrial: number;
    dropoutRatePerTrial: number;
    queryResolutionTime: string;
    protocolDeviationRate: number;
    complicationRate: number;
}

// Site Metrics
export interface SiteMetrics {
    screenFailureRate: number;
    avgTimeToContract: string;
    avgTimeToIRBApproval: string;
}

// IRB Info
export interface IRBInfo {
    name: string;
    registrationNumber: string;
    contactEmail: string;
}

// Staff Count
export interface StaffCount {
    doctor: number;
    technician: number;
}

// Financials (Site level)
export interface SiteFinancials {
    contractValueUSD: number;
    paymentsMadeUSD: number;
}

// Full Site Interface 
export interface Site {
    siteId: string;
    siteName: string;
    siteType: string;
    country: string;
    state: string;
    city: string;
    latitude: number;
    longitude: number;
    principalInvestigator: PrincipalInvestigator;
    investigatorExperience: number;
    contactInfo: ContactInfo;
    siteActivationDate: string;
    recruitmentCapacity: number;
    infrastructureScore: number;
    sitePerformance: SiteFullPerformance;
    siteMetrics: SiteMetrics;
    auditHistory: string[];
    performanceTrends: string;
    timelineDelays: Record<string, any>;
    IRBInfo: IRBInfo;
    patientRecruitmentSources: string[];
    equipment: string[];
    staffCount: StaffCount;
    monitoringVisits: number;
    financials: SiteFinancials;
    subjects: Subject[];
}

// Progress Metrics
export interface ProgressMetrics {
    enrolled: number;
    screened: number;
    completed: number;
    dropped: number;
    completionPercentage: number;
}

// Ethics Committee Approval
export interface EthicsCommitteeApproval {
    committeeName: string;
    approvalDate: string;
    approvalId: string;
}

// Full Protocol Interface 
export interface Protocol {
    protocolId: string;
    protocolName: string;
    nctId: string;
    drugName: string;
    therapeuticArea: string;
    phase: string;
    status: string;
    studyDesign: string;
    principalInvestigator: string;
    coordinatingCenter: string;
    budgetUSD: number;
    startDate: string;
    expectedEndDate: string;
    actualEndDate: string;
    estimatedEnrollmentSite: number;
    requiredSiteEnrollment: number;
    estimatedEnrollmentSubject: number;
    requiredSubjectEnrollment: number;
    protocolVersion: string;
    lastUpdatePosted: string;
    timelineDelays: Record<string, any>;
    progressMetrics: ProgressMetrics;
    dataMonitoringCommittee: string;
    ethicsCommitteeApproval: EthicsCommitteeApproval;
    regulatorySubmissions: string[];
    patientReportedOutcomes: string;
    sites: Site[];
}

// Nested interface for Financials 
export interface CompanyFinancials {
    revenueUSD: number;
    rdExpenditureUSD: number;
    marketCapUSD: number;
}

// Nested interface for Pipeline Summary
export interface PipelineSummary {
    phase1Count: number;
    phase2Count: number;
    phase3Count: number;
}

// Full Company Interface 
export interface Company {
    companyId: string;
    companyName: string;
    sponsorType: string;
    financials: CompanyFinancials;
    pipelineSummary: PipelineSummary;
    activeRegions: string[];
    therapeuticAreasCovered: string[];
    complianceScore: number;
    riskLevel: string;
    website: string;
    foundedYear: number;
    trialsRan: number;
    successRatio: number;
    avgTrialCompletionTime: number;
    totalSubjectsHandled: number;
    protocols: Protocol[];
}